### PR TITLE
Fixed longitude wrap

### LINKED ISSUE #1246: Longitude wrap when crossing the bearing straight
When we calculate distance between points, we do not take into account crossing over the 180 deg longitude barrier (line between Russia and Alaska).  This appears both in the AP_Common lib and in arducopter's position_vector.pde.

### PR DESCRIPTION
**Pull request by tridge**
This allows us to fly across the line at +/- 180 degrees longitude
fixes #1246
This is crying out for an auto-test
Note that we also need to fix ground stations. Both mavproxy and MissionPlanner can't display maps that cross the boundary
MAVProxy map fix here: https://github.com/ArduPilot/MAVProxy/pull/899

tested with mission around Rabi:
image

### CODE REVIEW

**Comment by peterbarker** [comment]
Branch with a unit test for this is here: https://github.com/peterbarker/ardupilot/tree/pr/lng-wrap-tests

Test fails on master but passes on this branch.

**Comment by rmackay9** [comment]
This is really fantastic.  I hope that at some point someone will build a boat frame capable of crossing the Atlantic with AP as the autopilot.

**Comment by tridge** [comment]
> Branch with a unit test for this is here:

thanks! added to this PR

**Review by khancyr — COMMENTED** [review]
that seems good
there still one place that should benefit from this : https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/commands_logic.cpp#L595 and

File: libraries/AP_Common/Location.cpp [hunk 1]
```diff
@@ -194,7 +194,7 @@ bool Location::get_vector_xy_from_origin_NE(Vector2f &vec_ne) const
         return false;
     }
     vec_ne.x = (lat-ekf_origin.lat) * LATLON_TO_CM;
```
Comment (durka) [inline libraries/AP_Common/Location.cpp:196]: If you are flying near the poles isn't it hazardous to subtract latitudes as well?

Comment (tridge) [reply libraries/AP_Common/Location.cpp:196]: no, latitude doesn't wrap, it just hits a limit of 90. We do have issues at the poles which will take a separate PR to fix, but it isn't the same issue as longitude wrap

**Comment by tridge** [comment]
> there still one place that should benefit from this

thanks!

File: libraries/AP_Common/Location.cpp (hunk 1 above)
Comment (durka) [reply libraries/AP_Common/Location.cpp:196]: It doesn't wrap but if you fly right over the pole from a latitude of +89 to +89, the velocity shouldn't be zero right?

Comment (durka) [reply libraries/AP_Common/Location.cpp:196]: Maybe that's what you meant by other issues.

**Comment by rmackay9** [comment]
Looks like AP_Spektrum_Telem::calc_gps_location() has an issue too ([here](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp#L492)) with a line like this:

> const uint32_t ulng = roundf((int32_t(lng) * 100.0f + (lng - int32_t(lng)) * 60.0f) * 10000.0f);

**Comment by rmackay9** [comment]
I suspect we should also fixup this one

- Tools/autotest/common.py's get_bearing ([see code](https://github.com/ArduPilot/ardupilot/blob/master/Tools/autotest/common.py#L4457))

**Comment by tridge** [comment]
> const uint32_t ulng = roundf((int32_t(lng) * 100.0f + (lng - int32_t(lng)) * 60.0f) * 10000.0f);

I don't think it needs the same change. That is about rounding, not difference between two longitudes

**Comment by tridge** [comment]
i've found an EKF issue if you travel more than half way around the world. The position innovation gets huge when you reach 180 degrees longitude from the takeoff point. I'll likely need help from @priseborough to fix

**Comment by priseborough** [comment]
> i've found an EKF issue if you travel more than half way around the world. The position innovation gets huge when you reach 180 degrees longitude from the takeoff point. I'll likely need help from @priseborough to fix
> Note that this should not hold up this PR.

We can move the EKF origin periodically to the match the current position.

**Review by peterbarker — APPROVED** [review]

**Pull request merged**